rbd & cephfs: modify fencing logic

Rakshith-R opened the pull request:
# Describe what this PR does #

This commit does the following:
- auto fence duration will be 1 year
- always remove client address metadata in ControllerUnpublish()
- removes logic to unfence in ControllerPublish()
- add logic to auto unfence in GetFenceClient()
  - if client itself is blocked and expiry is within auto fence duration of 1 yr
  - and after cool down period of 5 mins
  - client will auto un-fence itself.

# Testing logs #

- After mounting pod:
```
bash-5.1$ rbd image-meta ls ocs-storagecluster-cephblockpool/csi-vol-7e184c3c-6cff-48e8-a51f-058cf85
40fd9
There are 6 metadata on this image:

Key                                                                                                                          Value                                                         
.rbd.csi.ceph.com/clientaddress/0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9/compute-0  100.64.0.5  
...
```

```
bash-5.1$ ceph fs subvolume metadata ls ocs-storagecluster-cephfilesystem csi-vol-5d0b6c0e-380c-4388-b5f8-75181989b519 csi
{
    ".cephfs.csi.ceph.com/clientaddress/0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519/compute-0": "100.64.0.7",
    ".cephfs.csi.ceph.com/userid/0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519/compute-0": "csi-cephfs-node-ceph-user-g1-cd643577-babc-42cd-9fe4-2b8028f6ad18",
    "csi.ceph.com/cluster/name": "ocs-storagecluster-cephcluster",
    "csi.storage.k8s.io/pv/name": "pvc-a4d94eeb-a664-4daf-888c-50c1c3e3833e",
    "csi.storage.k8s.io/pvc/name": "cephfs-pvc",
    "csi.storage.k8s.io/pvc/namespace": "openshift-storage"
}
```


- after tainting the node
```
bash-5.1$ ceph osd blocklist ls
listed 0 entries
bash-5.1$ ceph osd blocklist ls
cidr:100.64.0.5:0/32 2026-12-04T16:39:40.177691+0000
listed 1 entries
bash-5.1$ rbd image-meta ls ocs-storagecluster-cephblockpool/csi-vol-7e184c3c-6cff-48e8-a51f-058cf8540fd9
There are 4 metadata on this image:

Key                               Value                                   
csi.ceph.com/cluster/name         ocs-storagecluster-cephcluster          
csi.storage.k8s.io/pv/name        pvc-ef585b38-2d67-4651-8223-7021246d3f3a
csi.storage.k8s.io/pvc/name       raw-block-rbd-pvc                       
csi.storage.k8s.io/pvc/namespace  openshift-storage    
```

```
bash-5.1$ ceph osd blocklist ls
10.131.0.27:0/4045147051 2025-12-12T07:45:40.759269+0000
...
10.131.0.27:6801/1650798790 2025-12-12T06:35:54.795175+0000
listed 21 entries

```
bash-5.1$ ceph osd blocklist ls
10.131.0.27:0/3058396927 2025-12-12T06:33:01.511355+0000
...
10.131.0.27:0/423785367 2025-12-12T07:45:40.759269+0000
cidr:100.64.0.7:0/32 2026-12-11T20:57:21.075994+0000
listed 22 entries
bash-5.1$ ceph fs subvolume metadata ls ocs-storagecluster-cephfilesystem csi-vol-5d0b6c0e-380c-4388-b5f8-75181989b519 csi
{
    "csi.storage.k8s.io/pv/name": "pvc-a4d94eeb-a664-4daf-888c-50c1c3e3833e",
    "csi.storage.k8s.io/pvc/name": "cephfs-pvc",
    "csi.storage.k8s.io/pvc/namespace": "openshift-storage"
}
```
```

- ControllerUnpublish:
```
I1204 10:50:27.736913       1 utils.go:350] ID: 24 Req-ID: 0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I1204 10:50:27.736980       1 utils.go:351] ID: 24 Req-ID: 0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9 GRPC request: {"node_id":"compute-0","volume_id":"0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9"}
I1204 10:50:27.739439       1 reflector.go:358] "Starting reflector" type="*v1.Secret" resyncPeriod="10m0s" reflector="github.com/ceph/ceph-csi/internal/util/k8s/secrets.go:107"
I1204 10:50:27.739467       1 reflector.go:404] "Listing and watching" type="*v1.Secret" reflector="github.com/ceph/ceph-csi/internal/util/k8s/secrets.go:107"
I1204 10:50:27.806817       1 omap.go:89] ID: 24 Req-ID: 0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.7e184c3c-6cff-48e8-a51f-058cf8540fd9"): map[csi.imageid:5fba6153f8a7 csi.imagename:csi-vol-7e184c3c-6cff-48e8-a51f-058cf8540fd9 csi.volname:pvc-ef585b38-2d67-4651-8223-7021246d3f3a csi.volume.owner:openshift-storage]
I1204 10:50:27.876431       1 reflector.go:436] "Caches populated" type="*v1.Secret" reflector="github.com/ceph/ceph-csi/internal/util/k8s/secrets.go:107"
I1204 10:50:28.110260       1 controllerserver.go:1790] ID: 24 Req-ID: 0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9 successfully removed user ID mapping for nodeId compute-0
I1204 10:50:29.041667       1 utils.go:357] ID: 24 Req-ID: 0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9 GRPC response: {}
```

```
I1211 15:07:57.286803       1 utils.go:350] ID: 31 Req-ID: 0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I1211 15:07:57.286954       1 utils.go:351] ID: 31 Req-ID: 0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519 GRPC request: {"node_id":"compute-0","volume_id":"0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519"}
I1211 15:07:57.292323       1 reflector.go:358] "Starting reflector" type="*v1.Secret" resyncPeriod="10m0s" reflector="github.com/ceph/ceph-csi/internal/util/k8s/secrets.go:107"
I1211 15:07:57.292365       1 reflector.go:404] "Listing and watching" type="*v1.Secret" reflector="github.com/ceph/ceph-csi/internal/util/k8s/secrets.go:107"
I1211 15:07:57.428723       1 reflector.go:436] "Caches populated" type="*v1.Secret" reflector="github.com/ceph/ceph-csi/internal/util/k8s/secrets.go:107"
I1211 15:08:08.377999       1 omap.go:89] ID: 31 Req-ID: 0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519 got omap values: (pool="ocs-storagecluster-cephfilesystem-metadata", namespace="csi", name="csi.volume.5d0b6c0e-380c-4388-b5f8-75181989b519"): map[csi.imagename:csi-vol-5d0b6c0e-380c-4388-b5f8-75181989b519 csi.volname:pvc-a4d94eeb-a664-4daf-888c-50c1c3e3833e csi.volume.owner:openshift-storage]
I1211 15:08:09.015404       1 controllerserver.go:1281] ID: 31 Req-ID: 0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519 userID mapping metadata .cephfs.csi.ceph.com/userid/0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519/compute-0 unset for subvolume csi-vol-5d0b6c0e-380c-4388-b5f8-75181989b519
I1211 15:08:10.037632       1 utils.go:357] ID: 31 Req-ID: 0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519 GRPC response: {}
```

- After removing taint and nodeplugin comes back on the node:
```
I1204 10:54:35.071053 1519008 utils.go:350] ID: 6 GRPC call: /fence.FenceController/GetFenceClients
I1204 10:54:35.071313 1519008 utils.go:351] ID: 6 GRPC request: {"parameters":{"clusterID":"openshift-storage"},"secrets":"***stripped***"}
I1204 10:54:35.125143 1519008 network_fence.go:230] ID: 6 auto-unfencing client with address "100.64.0.5/32"
I1204 10:54:36.190540 1519008 utils.go:357] ID: 6 GRPC response: {"clients":[{"addresses":[{"cidr":"100.64.0.5/32"}],"id":"ece6d53d-44cf-4362-9b1f-fe3da2d7966b"}]}
```

- blocklist and image meta after mounting pod again
```
bash-5.1$ ceph osd blocklist ls
listed 0 entries
---

bash-5.1$ rbd image-meta ls ocs-storagecluster-cephblockpool/csi-vol-7e184c3c-6cff-48e8-a51f-058cf8540fd9
There are 6 metadata on this image:

Key                                                                                                                          Value                                                         
.rbd.csi.ceph.com/clientaddress/0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9/compute-0  100.64.0.5/32                                                 
.rbd.csi.ceph.com/userid/0001-0011-openshift-storage-0000000000000003-7e184c3c-6cff-48e8-a51f-058cf8540fd9/compute-0         csi-rbd-node-ceph-user-g1-ed824a5e-04c9-4e20-a9d8-29f0af9f5182
csi.ceph.com/cluster/name                                                                                                    ocs-storagecluster-cephcluster                                
csi.storage.k8s.io/pv/name                                                                                                   pvc-ef585b38-2d67-4651-8223-7021246d3f3a                      
csi.storage.k8s.io/pvc/name                                                                                                  raw-block-rbd-pvc                                             
csi.storage.k8s.io/pvc/namespace                                                                                             openshift-storage 
```

```
bash-5.1$ ceph fs subvolume metadata ls ocs-storagecluster-cephfilesystem csi-vol-5d0b6c0e-380c-4388-b5f8-75181989b519 csi
{
    ".cephfs.csi.ceph.com/clientaddress/0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519/compute-0": "100.64.0.7",
    ".cephfs.csi.ceph.com/userid/0001-0011-openshift-storage-0000000000000001-5d0b6c0e-380c-4388-b5f8-75181989b519/compute-0": "csi-cephfs-node-ceph-user-g1-cd643577-babc-42cd-9fe4-2b8028f6ad18",
    "csi.ceph.com/cluster/name": "ocs-storagecluster-cephcluster",
    "csi.storage.k8s.io/pv/name": "pvc-a4d94eeb-a664-4daf-888c-50c1c3e3833e",
    "csi.storage.k8s.io/pvc/name": "cephfs-pvc",
    "csi.storage.k8s.io/pvc/namespace": "openshift-storage"
}
```

- cooldown period error

```
I1216 11:05:49.238891       1 utils.go:350] ID: 104 GRPC call: /fence.FenceController/GetFenceClients
I1216 11:05:49.238934       1 utils.go:351] ID: 104 GRPC request: {"parameters":{"clusterID":"openshift-storage"},"secrets":"***stripped***"}
E1216 11:05:49.240216       1 fencing.go:519] ID: 104 failed to auto unfence client: blocklist entry "100.64.0.6:0/32" is still in cool-down period for 2m57.000622843s
E1216 11:05:49.240255       1 utils.go:355] ID: 104 GRPC error: rpc error: code = Internal desc = failed to unfence client: blocklist entry "100.64.0.6:0/32" is still in cool-down period for 2m57.000622843s
```


---
**Checklist:**

* [ ] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [ ] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.


Depends-on: #5867 

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

Rakshith-R commented:
Added testing logs in pr description. 
Please take a look.

Madhu-1 commented on docs/design/proposals/non-graceful-node-shutdown.md:
```diff
@@ -29,7 +29,20 @@ of service.
 > or lingering mounts, which can lead to serious data integrity issues.
 
-## Proposed solution
+## Proposed solution 2
```
Lets keep the only working solution as the previous solution is not a valid solution anymore

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -178,5 +188,78 @@ func (fcs *FenceControllerServer) GetFenceClients(
 		},
 	}
 
+	if options[disableAutoUnfenceKey] != "false" {
```
IMO we dont need it we already have separate API to fence and unfence, lets keep it as it is

Madhu-1 commented on docs/design/proposals/non-graceful-node-shutdown.md:
```diff
@@ -29,7 +29,20 @@ of service.
+
+Removal of the client from blocklist can be automated by enabling CSIAddons
+in CephCSI deployment. CephCSI will auto unblocklist itself during CSIAddons
+`GetFenceClient()` call when pods are rescheduled after node recovery.
```
cant we do it during NodeStage?

Rakshith-R replied:
> cant we do it during NodeStage?

**unblocklist on nodestage will not work since nodestage may occur within a faulty node** 

GetFenceClient() request will only come from an external controller when pods start getting scheduled on recovered node back again.

Madhu-1 replied:
This will not work for cephfs PVC and it puts dependency on the csiaddons itself.

Rakshith-R replied:
> This will not work for cephfs PVC and it puts dependency on the csiaddons itself.

The fencing will work for cephfs PVC too.

Unfencing will be taken care by CSIAddons GetFenceClient Request which only RBD supports right now but
nodeplugins run on host network so both RBD and CephFS volumes will be able to mount again.

The whole feature is kind of experimental. We can extend cephFS support for GetFenceClient() call in next releases.

Madhu-1 replied:
in cephfs only clusters we will get into the problem with this feature.

Madhu-1 replied:
>unblocklist on nodestage will not work since nodestage may occur within a faulty node

Can we unblocklist if there is no taint on the node? and dont allow the proceed to Mount if the taint is present?

Rakshith-R replied:
> > unblocklist on nodestage will not work since nodestage may occur within a faulty node
> 
> Can we unblocklist if there is no taint on the node? and dont allow the proceed to Mount if the taint is present?

API server may or may not be responsive in such times / output may or may not be reliable. A race condition may occur where we get no taint response when at the same time user adds taint and blocklist happens while we go and unblocklist it.

Madhu-1 replied:
Lets have a concert answer when this can be possible, If API server is not responsive lets not let the NodeStage pass through, if the taints are added later i assume the node will get blocklisted again during the ControllerUnpublish call.

black-dragon74 commented on internal/cephfs/nodeserver.go:
```diff
@@ -421,6 +421,9 @@ func (ns *NodeServer) setClientAddress(
 	}
 
+	// adding /32 to the IP address to make it a CIDR block.
+	ipAddress += "/32"
```
The address might be v6 as well? `ParseClientIP` supports both.

black-dragon74 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -178,5 +188,76 @@ func (fcs *FenceControllerServer) GetFenceClients(
+		// blocklist list entry contains ':0' port in the address while
+		// go-ceph validation fails if we provide port in the addr.
+		// Therefore, use the modified address with ':0' only for comparison.
+		cephAddr := strings.TrimSuffix(addr, "/32") + ":0/32"
```
This will not work for IPv6.

black-dragon74 commented on internal/rbd/nodeserver.go:
```diff
@@ -505,6 +505,9 @@ func (ns *NodeServer) setClientAddress(
 	}
 
+	// adding /32 to the IP address to make it a CIDR block.
+	ipAddress += "/32"
```
Same concern around v6.

Rakshith-R committed:
rbd: modify fencing logic

This commit does the following:
- adds "ceph_preview" go build tag
  in order to use go-ceph osd admin pkg
- auto fence duration will be 1 year
- always remove client address metadata in
  ControllerUnpublish()
- removes logic to unfence in ControllerPublish()
- add logic to auto unfence in GetFenceClient()
  - if client itself is blocked and expiry is
    within auto fence duration of 1 yr
  - client will auto un fence itself.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
doc: modify design for non-graceful-node-shutdown

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
cephfs: modify fencing logic

This commit does the following:
- auto fence duration will be 1 year
- always remove client address metadata in
  ControllerUnpublish()
- removes logic to unfence in ControllerPublish()
- add support for GetFenceClient()

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
IPv6 and ceph's weird blocklist entry containing port ":0" in cidr range is handled now too.
With complete support for cephfs fencing.
(logs for cephfs fencing testing will be added to pr description)
PTAL

nixpanic commented on docs/design/proposals/non-graceful-node-shutdown.md:
```diff
@@ -45,33 +45,13 @@ This stored address can then be used:
-* In `ControllerPublishVolume()` to remove the client from the blocklist,
-  allowing the volume to be accessed again when the node recovers.
+Removal of blocklist can either be done manually or can be automated by enabling
+CSIAddons in CephCSI deployment and creating of networkFenceClass. CephCSI will
```
_CSI-Addons in the Ceph-CSI deployment and creation of a NetworkFenceClass_

Always spell **Ceph-CSI** and **CSI-Addons** like this.

nixpanic commented on Makefile:
```diff
@@ -53,7 +53,7 @@ GO_PROJECT=github.com/ceph/ceph-csi
 CEPH_VERSION ?= $(shell . $(CURDIR)/build.env ; echo $${CEPH_VERSION})
 # The 'ceph_preview' tag may be needed for unstable/unreleased APIs.
-GO_TAGS_LIST ?= $(CEPH_VERSION)
+GO_TAGS_LIST ?= $(CEPH_VERSION) ceph_preview
```
At least the commit message should mention why this is needed, which APIs? It makes it easier to review when we can drop the `ceph_preview` tag again.

Rakshith-R replied:
its for go-ceph osd admin pkg

added a line to the comment

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -452,3 +456,168 @@ func (nf *NetworkFence) parseBlocklistForCIDR(ctx context.Context, blocklist, ci
+		return nil, status.Errorf(codes.Internal, "failed to convert IP to CIDR: %s", err)
+	}
+
+	resp := &fence.GetFenceClientsResponse{
```
please move the creation of `resp` after calling `autoUnfenceClientOnMatch()`

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -452,3 +456,168 @@ func (nf *NetworkFence) parseBlocklistForCIDR(ctx context.Context, blocklist, ci
+		}
+
+		// Until is in ISO8601 format.
+		until, err := time.Parse(ISO8601TimeLayout, entry.Until)
```
Should go-ceph not have `osdAdmin.Blocklist.Until` as a `time.Time` type? At the very least go-ceph should have an option to convert the value to `time.Time`.

@anoopcs9, what do you think?

anoopcs9 replied:
I've created https://github.com/ceph/go-ceph/pull/1209. Let me know if that's sufficient.

Rakshith-R replied:
Thanks !

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1687,69 +1688,10 @@ func (cs *ControllerServer) ControllerExpandVolume(
 }
 
 // ControllerPublishVolume implements the CSI ControllerPublishVolume RPC.
```
This comment isn't really correct anymore either. It does not implement anything, just returns a success response. Please update the comment with an explanation why this is needed and can not be removed.

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -452,3 +456,168 @@ func (nf *NetworkFence) parseBlocklistForCIDR(ctx context.Context, blocklist, ci
+}
+
+// matchEntry checks if the entry address matches the provided address while
+// suffix (:0/32 for IPv4 and :0/128 for IPv6) is ignored accordingly.
```
where does the `:0/32` or `:0/128` suffix come from? CIDR format would just have `/32` or `/128` as suffix. Can you include an example of the `actual` parameter that is passed?

Is there a guarantee that the `0` in the suffix (Ceph clien session?) is always `0`?

Rakshith-R replied:
> where does the `:0/32` or `:0/128` suffix come from? CIDR format would just have `/32` or `/128` as suffix. Can you include an example of the `actual` parameter that is passed?
> 
> Is there a guarantee that the `0` in the suffix (Ceph clien session?) is always `0`?

I had a comment regarding this but it seems to be removed while making changes to accommodate ipv6.

`ceph osd blocklist ls` lists the address with `":0"` even though the input is a cidr range with no port, such format is invalid and `ceph osd blocklist range rm` itself command will not work when ":0" is used.

I'll re-add the comment explaining this.

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -452,3 +456,175 @@ func (nf *NetworkFence) parseBlocklistForCIDR(ctx context.Context, blocklist, ci
+// autoUnfenceClientOnMatch removes the client address from the blocklist
+// if it matches an entry with 'until' time less than or equal to
+// AutoBlocklistTime duration.
+func autoUnfenceClientOnMatch(
```
what happens if the csi controller unpublish blocklists the IP and the daemonset pod is not dead yet or it might be running and it received a RPC to get the NodeFence? daemonset pods are the one that gets evicted at last due to priorityclass?

Rakshith-R replied:
> what happens if the csi controller unpublish blocklists the IP and the daemonset pod is not dead yet or it might be running and it received a RPC to get the NodeFence? daemonset pods are the one that gets evicted at last due to priorityclass?

https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/

All the pods are forcefully deleted immediately.

Madhu-1 replied:
>All the pods are forcefully deleted immediately.

i slight delay will get back to the same problem its not 100% but there are chances of hitting it.

I still see two problems with this one.

1. Adds the dependency on the csiaddons for a feature
2. There are chances that we will get into the problem i explained about.

considering the above, what are the advantages that we are trying to get doing it at the csiaddons RPC calls rather than doing at the NodeStage level?

Rakshith-R replied:
> > All the pods are forcefully deleted immediately.
> 
> i slight delay will get back to the same problem its not 100% but there are chances of hitting it.
>
> I still see two problems with this one.
> 
> 1. Adds the dependency on the csiaddons for a feature
> 2. There are chances that we will get into the problem i explained about.
> 
> considering the above, what are the advantages that we are trying to get doing it at the csiaddons RPC calls rather than doing at the NodeStage level?


>> Lets have a concert answer when this can be possible, If API server is not responsive lets not let the NodeStage pass through, if the taints are added later i assume the node will get blocklisted again during the ControllerUnpublish call.

Regarding 1:

NodeStage is a frequent, performance sensitive operation, adding api server call dependency for each call is even worse compared to depending on csi-addons and strictly a "no" IMO.

The whole point of csiaddons is to offloads operations that cannot be accommodated into CSI framework.


Regarding 2:

> i slight delay will get back to the same problem its not 100% but there are chances of hitting it.

taint -> immediate pod deletion -> VA deletion -> controllerUnpublish

Nope, ControllerUnpublish can come only after pods are deleted. I don't this problem occuring.

Madhu-1 replied:
> > > All the pods are forcefully deleted immediately.
> > 
> > 
> > i slight delay will get back to the same problem its not 100% but there are chances of hitting it.
> > I still see two problems with this one.
> > 
> > 1. Adds the dependency on the csiaddons for a feature
> > 2. There are chances that we will get into the problem i explained about.
> > 
> > considering the above, what are the advantages that we are trying to get doing it at the csiaddons RPC calls rather than doing at the NodeStage level?
> 
> > > Lets have a concert answer when this can be possible, If API server is not responsive lets not let the NodeStage pass through, if the taints are added later i assume the node will get blocklisted again during the ControllerUnpublish call.
> 
> Regarding 1:
> 
> NodeStage is a frequent, performance sensitive operation, adding api server call dependency for each call is even worse compared to depending on csi-addons and strictly a "no" IMO.
> 
> The whole point of csiaddons is to offloads operations that cannot be accommodated into CSI framework.
> 

I dont see where we are making API call in this one, and also why do we required API call? if the API call is problem, you can have watchers etc

> Regarding 2:
> 
> > i slight delay will get back to the same problem its not 100% but there are chances of hitting it.
> 
> taint -> immediate pod deletion -> VA deletion -> controllerUnpublish
> 
> Nope, ControllerUnpublish can come only after pods are deleted. I don't this problem occuring.

There are two different pods here the one is using PVC and another is csiaddons and the cephcsi daemonset. If the application pods gets deleted first and later the csiaddons and csi daemonset there are chances the ControllerUnpublish is done before the csiaddon pod is deleted (nothing is guaranteed  that csiaddons will get deleted first before the last ControllerUnpublish)

Rakshith-R replied:
> > Regarding 2:
> > > i slight delay will get back to the same problem its not 100% but there are chances of hitting it.
> > 
> > 
> > taint -> immediate pod deletion -> VA deletion -> controllerUnpublish
> > Nope, ControllerUnpublish can come only after pods are deleted. I don't this problem occuring.
> 
> There are two different pods here the one is using PVC and another is csiaddons and the cephcsi daemonset. If the application pods gets deleted first and later the csiaddons and csi daemonset there are chances the ControllerUnpublish is done before the csiaddon pod is deleted (nothing is guaranteed that csiaddons will get deleted first before the last ControllerUnpublish)

Added a cool down of 5 minutes, CephCSI will not auto fence for the first 5 minutes after from the time auto fence is done.
This should take care of the above concern.
PTAL

Rakshith-R replied:
Added cooldown failure logs to pr.

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -452,3 +456,175 @@ func (nf *NetworkFence) parseBlocklistForCIDR(ctx context.Context, blocklist, ci
+
+// matchEntry checks if the actual address matches the expected address along
+// with the matching suffix (":0/32" for IPv4 and ":0/128" for IPv6).
+func matchEntry(actual, expected string) bool {
```
Will this also considers the IP's blocklisted by user or by ceph itself due to some network? Have we tested this case as well?

Rakshith-R replied:
> Will this also considers the IP's blocklisted by user or by ceph itself due to some network? Have we tested this case as well?

Ceph always blocks a particular client ip with nonce.
If a user blocklists the very same single `IP with CIDR fixing the entire IP for duration less than 1 year`, Then cephcsi will unblocklist it as is designed to do so.

Rakshith-R committed:
util: gatekeep autoUnfence in GetFenceClient req

This commit modifies code to make use of enableFence
config to determine whether to auto unfence or not.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
/test ci/centos/mini-e2e/k8s-1.34

Rakshith-R commented:
/test ci/centos/mini-e2e/k8s-1.34

Rakshith-R commented:
/test ci/centos/mini-e2e/k8s-1.34

Rakshith-R committed:
util: add a blocklist cooldown period of 5 mins

This commit adds a blocklist cool period of 5 mins
making sure cephcsi will wait for atleast 5 mins
before auto unfencing a node.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
/test ci/centos/mini-e2e/k8s-1.34

Madhu-1 commented on docs/design/proposals/non-graceful-node-shutdown.md:
```diff
@@ -45,33 +45,13 @@ This stored address can then be used:
-  allowing the volume to be accessed again when the node recovers.
+Removal of blocklist can either be done manually or can be automated by enabling
+CSI-Addons in Ceph-CSI deployment and creating of networkFenceClass. Ceph-CSI will
+auto unblocklist itself during CSI-Addons `GetFenceClient()` call when pods are
```
It would be nice to not tie this to `GetFenceClient` and introduce a new RPC

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -26,15 +26,22 @@ import (
+	// BlockListCoolDownPeriod defines the time duration
+	// after which a blocklist entry can be removed.
+	// TODO: Make this configurable.
+	blockListCoolDownPeriod = 5 * time.Minute
```
make it 20 or 30 minutes minutes to be on safer side?

Rakshith-R replied:
> make it 20 or 30 minutes minutes to be on safer side?

5 minutes cool down is only to avoid the following situation discussed above

> There are two different pods here the one is using PVC and another is csiaddons and the cephcsi daemonset. If the application pods gets deleted first and later the csiaddons and csi daemonset there are chances the ControllerUnpublish is done before the csiaddon pod is deleted (nothing is guaranteed that csiaddons will get deleted first before the last ControllerUnpublish)

After immediate garbage collection of pods on out of service tainted  node, there's no need to be concerned about GetFenceClients since there will be no pod in the cluster for csiaddons(to get ip from and) to send request to.

I think adding 20 minutes cool down will keep applications from coming up for too long if node is restored.

Rakshith-R replied:
Pod garbage collection on tainted out of service node is done in parallel and should be immediate

https://github.com/kubernetes/kubernetes/pull/108486/files#diff-eaebfd5acb747dd3dd13eb812a745b0a886ba2a24f82e852772c45ac0d936f9f

```
	klog.V(4).Infof("Garbage collecting %v pods that are terminating on node tainted with node.kubernetes.io/out-of-service", deleteCount)
	// sort only when necessary
	sort.Sort(byCreationTimestamp(terminatingPods))
	var wait sync.WaitGroup
	for i := 0; i < deleteCount; i++ {
		wait.Add(1)
		go func(namespace string, name string) {
			defer wait.Done()
			if err := gcc.deletePod(namespace, name); err != nil {
				// ignore not founds
				utilruntime.HandleError(err)
			}
		}(terminatingPods[i].Namespace, terminatingPods[i].Name)
	}
	wait.Wait()
}
```

Reboot may take 1-2 minutes to 15 mins.

https://docs.redhat.com/en/documentation/openshift_container_platform/4.12/html/updating_clusters/understanding-openshift-update-duration#estimating-cluster-update-time_openshift-update-duration

```
Note
The time it takes to reboot a particular node varies significantly. In cloud instances, the reboot might take about 1 to 2 minutes, whereas in physical bare metal hosts the reboot might take more than 15 minutes.
```

I think 5 minutes is a sweet spot between preventing immediate unfencing and quick resumption of ops after node recovery.

let me know what you think.

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -452,3 +459,188 @@ func (nf *NetworkFence) parseBlocklistForCIDR(ctx context.Context, blocklist, ci
 	return matchingHosts
 }
+
+// GetFenceClients fetches the ceph cluster ID and the client address that need to be fenced.
```
update the comment as its doing auto fencing as well

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -452,3 +459,188 @@ func (nf *NetworkFence) parseBlocklistForCIDR(ctx context.Context, blocklist, ci
+
+	fsID, err := conn.GetFSID()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get cephfs id: %s", err)
```
`failed to get ceph id`

Madhu-1 commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -256,3 +260,233 @@ listed 1 entries`,
+			args: args{
+				blocklist: &[]osdAdmin.Blocklist{
+					{
+						Addr:  "192.0.1.0:0/32",
```
add test for IPV6 as well

Madhu-1 commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -256,3 +260,233 @@ listed 1 entries`,
+	t.Parallel()
+	tests := []struct {
+		name string // description of this test case
+		// Named input parameters for target function.
```
remove the comments?

Madhu-1 commented on internal/util/cephcmds.go:
```diff
@@ -33,8 +33,9 @@ import (
-	InvalidPoolID int64 = -1
-	blocklistTime       = "157784760"
+	InvalidPoolID     int64 = -1
+	AutoBlocklistTime       = time.Second * 31556952  // 1 year in seconds
```
keep it for 3 years to be on safer side

iPraveenParihar commented on docs/design/proposals/non-graceful-node-shutdown.md:
````diff
@@ -121,6 +101,17 @@ This stored address can then be used:
         ceph osd blocklist add <clientAddress> 157784760
         ```
 
+   * Remove the `.[rbd|cephfs].csi.ceph.com/clientAddress/<NodeId>` metadata
````
please update everywhere with proper format?
```
.[rbd|cephfs].csi.ceph.com/clientAddress/<VolumeID>/<NodeID>
```

Rakshith-R replied:
> please update everywhere with proper format?
> 
> ```
> .[rbd|cephfs].csi.ceph.com/clientAddress/<VolumeID>/<NodeID>
> ```

unrelated to changes in this pr, can be done in a follow up pr.
This pr is already very big.

iPraveenParihar reviewed (approved):
LGTM. Thanks!

iPraveenParihar commented:
@Mergifyio rebase

iPraveenParihar commented:
@Mergifyio queue

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
@mergifyio queue

nixpanic commented:
@mergifyio requeue

nixpanic commented:
@mergifyio queue

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
@mergifyio queue

nixpanic commented:
@mergifyio requeue

There have been a few infrastructure issues in the CentOS CI. Hopefully these have been addressed now.

nixpanic commented:
@mergifyio queue

nixpanic commented:
@mergifyio requeue

nixpanic commented:
@mergifyio queue

iPraveenParihar commented:
@mergifyio requeue

iPraveenParihar commented:
@Mergifyio queue
